### PR TITLE
Implement QK attention head chunking for CSA [Deepseek v4]

### PR DESCRIPTION
**Pull request by octatrifan**
# Description

Stacked on [PR 5033](https://github.com/AI-Hypercomputer/maxtext/pull/5033) / `octatrifan-dsv4-indexer-loss`.
Dense attention scoring materializes full interaction tensors across all heads.
This causes OOM issues at longer context lengths.
PR [#4564](https://github.com/AI-Hypercomputer/maxtext/pull/4564) addressed this for MLA by chunking over the heads dimension with `jax.lax.scan`. This PR extends that optimization to DeepSeek-V4 Compressed Sparse Attention (CSA).

BUGS: b/556431288

# Changes
Chunking is applied to two sites using the dedicated `csa_qk_head_chunk_size` flag:
1. **`DeepseekV4Indexer`**: Scans over `index_n_heads // head_chunk_size` using `jax.checkpoint` on the scan body so backward activation memory stays bounded to 1 chunk.
2. **`calculate_csa_indexer_loss`**: Scans over `heads // head_chunk_size`. Since teacher inputs use `stop_gradient`, `jax.checkpoint` is omitted (`jax.lax.scan` lowers to an HLO while-loop boundary in XLA, preventing downstream fusion).
Defaults to `0` (native unchunked path).

Additional improvements:
- **Flag validation**: Added divisibility validation in `types.py` ensuring `csa_qk_head_chunk_size` cleanly divides `num_query_heads` and `indexer_n_heads`.
- **Direct einsum**: Optimized unchunked indexer scoring via `"bhsd, bwd -> bhsw"`, eliminating explicit 4D `compressed_kv` tensor broadcasting.
- **Mask cleanup**: Removed redundant 4D intermediate token masking in chunked and unchunked loss paths.

# Tests
Added 4 unit tests in `tests/unit/deepseek_v4_indexer_loss_test.py`:
- `test_csa_indexer_scoring_head_chunking_parity`: Verifies mathematical parity between unchunked ($c=0$) and chunked ($c=2$) forward scores and top-k routing indices.
- `test_csa_indexer_loss_head_chunking_parity`: Verifies mathematical parity between unchunked and chunked teacher CSA loss across multiple chunk sizes.
- `test_csa_indexer_chunked_gradients_flow`: Verifies gradient flow through the chunked scan body into indexer projections without gradient leakage into main model components.
- `test_csa_indexer_gradients_chunking_parity`: Verifies parameter gradients (`q_proj`, `kv_proj`, `gate_proj`, `weights_proj`) match between chunked and non-chunked execution within FP tolerance.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/attention_compressed.py [hunk 1]
```diff
@@ -1859,15 +1891,37 @@ def calculate_csa_indexer_loss(
...
+    else:
+      attention_scores = jnp.einsum("bthd, bwd -> bhtw", query, k_vec, precision=self.config.matmul_precision)
+      attention_scores = attention_scores + c_mask
+
+      # Apply NaN shielding for pre-block tokens
+      safe_scores = jnp.where(valid_tokens_mask[:, None, :, None], attention_scores, 0.0)
+      raw_probs = jax.nn.softmax(safe_scores.astype(jnp.float32), axis=-1)
+      raw_probs = jnp.where(valid_tokens_mask[:, None, :, None], raw_probs, 0.0)
+      target_probs = jnp.sum(raw_probs, axis=1)
+      target_probs = jax.lax.optimization_barrier(target_probs)
```
Comment (dandragona) [inline src/maxtext/layers/attention_compressed.py:1924]: Do we need this in the chunking branch too? Tbh not sure.

File: src/maxtext/configs/base.yml [hunk 2]
```diff
@@ -204,7 +204,7 @@ logits_dot_in_fp32: false  # whether to use fp32 in logits_dense or shared_embed
 cast_logits_to_fp32: true # whether to cast the logits to fp32. the higher precision is generally beneficial, but it can vary slightly.
 float32_qk_product: false # in dot_product attention, whether to cast to fp32 the inputs to qk product
 float32_logits: false # in dot_product attention, whether to cast to fp32 the inputs to softmax
-mla_qk_head_chunk_size: 0 # Limits HBM footprint by sequentially evaluating the QK matrix in the Indexer across the unsharded local heads dimension natively.
+mla_qk_head_chunk_size: 0 # Limits HBM footprint by sequentially evaluating the QK matrix in MLA and CSA Indexers across the unsharded local heads dimension natively.
```
Comment (dandragona) [inline src/maxtext/configs/base.yml]: Might already exist, but raise an error when we validate flags to ensure this divides head size.

Comment (dandragona) [inline src/maxtext/configs/base.yml]: Also I think we have overloaded this flag as it is for MLA originally and now its name is confusing.

I think you should rename it and update: 
google3/cloud/cluster/supercomputer/ubench/core/workloads/maxtext_training/configs/tpu7x-4x8x8/deepseekv3.2_671b_128K.yaml

Or just split to its own flag.

File: tests/unit/deepseek_v4_indexer_loss_test.py [hunk 3]
```diff
@@ -203,6 +206,123 @@ def test_csa_indexer_loss_kl_divergence_zero(self):
...
+        hidden_states=hidden_states,
+        q_latent=q_latent,
+        position_ids=positions,
+        model_mode=MODEL_MODE_TRAIN,
+        return_scores=True,
+    )
+    np.testing.assert_array_equal(np.array(topk_chunked), np.array(topk_native))
+    np.testing.assert_allclose(np.array(scores_chunked), np.array(scores_native), rtol=1e-5, atol=1e-5)
+
+  def test_csa_indexer_chunked_gradients_flow(self):
```
Comment (dandragona) [inline tests/unit/deepseek_v4_indexer_loss_test.py:282]: I don't think we are testing whether the chunked gradients are equal to the non-chunked gradients.

**Review by dandragona — COMMENTED** [review]
Mostly minor comments.

File: src/maxtext/layers/attention_compressed.py (hunk 1 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py:1924]: Not needed in the chunking branch. `jax.lax.scan` emits an explicit HLO while-loop boundary in XLA, which prevents downstream compiler fusion across it anyway. Kept the barrier only in the non-chunked path where fusion can occur.

File: src/maxtext/configs/base.yml (hunk 2 above)
Comment (octatrifan) [reply src/maxtext/configs/base.yml]: Added validation in types.py verifying that the chunk size evenly divides num_query_heads and indexer_n_heads.

Comment (octatrifan) [reply src/maxtext/configs/base.yml]: Split it into a separate `csa_qk_head_chunk_size` flag so `mla_qk_head_chunk_size` and V3.2 remain untouched.

File: tests/unit/deepseek_v4_indexer_loss_test.py (hunk 3 above)
Comment (octatrifan) [reply tests/unit/deepseek_v4_indexer_loss_test.py:282]: Added `test_csa_indexer_gradients_chunking_parity` to verify that parameter gradients between chunked and non-chunked runs match (within numerical tolerance).

File: src/maxtext/configs/base.yml [hunk 4]
```diff
@@ -206,6 +206,7 @@ cast_logits_to_fp32: true # whether to cast the logits to fp32. the higher preci
 float32_qk_product: false # in dot_product attention, whether to cast to fp32 the inputs to qk product
 float32_logits: false # in dot_product attention, whether to cast to fp32 the inputs to softmax
 mla_qk_head_chunk_size: 0 # Limits HBM footprint by sequentially evaluating the QK matrix in the Indexer across the unsharded local heads dimension natively.
+csa_qk_head_chunk_size: 0 # Limits HBM footprint by sequentially evaluating the QK matrix in the CSA Indexer across the unsharded local heads dimension natively.
```
Comment (shuningjin) [inline src/maxtext/configs/base.yml]: Can you add more comments to explain the difference between the two?
- is `mla_qk_head_chunk_size` use for ds32? and ds3?
- and `csa_qk_head_chunk_size` use for ds4?

File: src/maxtext/layers/attention_compressed.py [hunk 5]
```diff
@@ -883,20 +883,52 @@ def indexer_compressor_fn(buf_kv, buf_gate):
...
+    else:
+      compressed_kv = jnp.expand_dims(compressed, axis=1)
+      compressed_kv = jnp.broadcast_to(
+          compressed_kv, (batch_size, self.index_n_heads, compressed_len, self.index_head_dim)
+      )
+
+      q = q.astype(jnp.float32)
+      compressed_kv = compressed_kv.astype(jnp.float32)
+
+      scores = jnp.einsum("bhsd,bhwd->bhsw", q, compressed_kv, precision=self.config.matmul_precision)
```
Comment (shuningjin) [inline src/maxtext/layers/attention_compressed.py]: Similar to the chunked path, can we also eliminate `compressed_kv` here? Instead, directly use
```
scores = jnp.einsum(
    "bhsd, bwd -> bhsw",
    q.astype(jnp.float32),
    compressed.astype(jnp.float32),
    precision=self.config.matmul_precision,
)
```

File: src/maxtext/layers/attention_compressed.py [hunk 6]
```diff
@@ -1859,15 +1891,37 @@ def calculate_csa_indexer_loss(
...
+
+      def scan_body_heads(carry, xs):
+        q_c = xs["q"]  # [h_chunk, b, t, d]
+        attn_chunk = jnp.einsum("hbtd, bwd -> bhtw", q_c, k_vec, precision=self.config.matmul_precision)
+        attn_chunk = attn_chunk + c_mask
+
+        # Apply NaN shielding for pre-block tokens
+        safe_attn = jnp.where(valid_tokens_mask[:, None, :, None], attn_chunk, 0.0)
+        probs_chunk = jax.nn.softmax(safe_attn.astype(jnp.float32), axis=-1)
+        probs_chunk = jnp.where(valid_tokens_mask[:, None, :, None], probs_chunk, 0.0)
```
Comment (shuningjin) [inline src/maxtext/layers/attention_compressed.py]: can remove? we already have mask in the end
```
target_probs = jnp.where(valid_tokens_mask[:, :, None], target_probs, 0.0)
```

File: src/maxtext/layers/attention_compressed.py [hunk 7]
```diff
@@ -1859,15 +1891,37 @@ def calculate_csa_indexer_loss(
...
+      init_probs = jnp.zeros((batch, q_len, compressed_len), dtype=jnp.float32)
+      target_probs, _ = jax.lax.scan(scan_body_heads, init_probs, {"q": q_h})
+    else:
+      attention_scores = jnp.einsum("bthd, bwd -> bhtw", query, k_vec, precision=self.config.matmul_precision)
+      attention_scores = attention_scores + c_mask
+
+      # Apply NaN shielding for pre-block tokens
+      safe_scores = jnp.where(valid_tokens_mask[:, None, :, None], attention_scores, 0.0)
+      raw_probs = jax.nn.softmax(safe_scores.astype(jnp.float32), axis=-1)
+      raw_probs = jnp.where(valid_tokens_mask[:, None, :, None], raw_probs, 0.0)
```
Comment (shuningjin) [inline src/maxtext/layers/attention_compressed.py]: can remove? we already have mask in the end
```
target_probs = jnp.where(valid_tokens_mask[:, :, None], target_probs, 0.0)
```

File: src/maxtext/layers/attention_compressed.py (hunk 1 above)
Comment (shuningjin) [reply src/maxtext/layers/attention_compressed.py:1947]: would be good to add the explanation in code

**Review by shuningjin — COMMENTED** [review]
Overall is good. Some minor comments.

**Review by dandragona — APPROVED** [review]

File: src/maxtext/configs/base.yml (hunk 4 above)
Comment (octatrifan) [reply src/maxtext/configs/base.yml]: Updated the comments in `base.yml` to explicitly note that `mla_qk_head_chunk_size` is for DeepSeek-V3 / V3.2 (MLA) and `csa_qk_head_chunk_size` is for DeepSeek-V4 (CSA).

File: src/maxtext/layers/attention_compressed.py (hunk 5 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py]: Switched the unchunked path to eliminate the compressed_kv broadcast, just like indicated by the code block you shared.

File: src/maxtext/layers/attention_compressed.py (hunk 6 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py]: Removed the `jnp.where` masking in both chunked and unchunked paths since `target_probs` is masked right after the head sum.

File: src/maxtext/layers/attention_compressed.py (hunk 7 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py]: Removed the jnp.where masking in both chunked and unchunked paths since target_probs is masked right after the head sum.

File: src/maxtext/layers/attention_compressed.py (hunk 1 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py:1947]: Added inline comments above `optimization_barrier` explaining that `jax.lax.scan` emits an explicit HLO while-loop boundary in XLA, making the barrier redundant in the chunked branch.

**Review by shuningjin — APPROVED** [review]
Thanks for the thoughtful extension on optimization! LGTM

File: src/maxtext/layers/attention_compressed.py [hunk 8]
```diff
@@ -1875,15 +1904,37 @@ def calculate_csa_indexer_loss(
...
-
-    # Apply NaN shielding for pre-block tokens
-    safe_scores = jnp.where(valid_tokens_mask[:, None, :, None], attention_scores, 0.0)
-    raw_probs = jax.nn.softmax(safe_scores.astype(jnp.float32), axis=-1)
-    raw_probs = jnp.where(valid_tokens_mask[:, None, :, None], raw_probs, 0.0)
-    target_probs = jnp.sum(raw_probs, axis=1)
-    target_probs = jax.lax.optimization_barrier(target_probs)
+    head_chunk_size = getattr(self.config, "csa_qk_head_chunk_size", 0)
+    if head_chunk_size > 0:
+      num_chunks = heads // head_chunk_size
```
Comment (parambole) [inline src/maxtext/layers/attention_compressed.py:1917]: Nit: 

One flag, two divisors — worth a comment at each site.

This line divides `heads`, which is `query.shape[2]`, i.e. `num_query_heads`. Line 894 divides
`self.index_n_heads`. Both are correct and your validation covers both, which is why there are two
checks in `types.py` rather than one — but someone who hits this site first will reasonably assume
`csa_qk_head_chunk_size` is an indexer-head quantity, then be briefly confused about why the
validation talks about `num_query_heads`.

```suggestion
      # Teacher chunks the main model's query heads; the student path at line 894
      # chunks indexer heads. Both are validated in configs/types.py.
      num_chunks = heads // head_chunk_size
```

**Review by parambole — APPROVED** [review]
LGTM 🚀

File: src/maxtext/layers/attention_compressed.py (hunk 8 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py:1917]: Added the comment. Did not mention "line 894" as this file can change in the future and no longer point to the desired line, therefore I referenced the target method by name (DeepseekV4Indexer) instead.
